### PR TITLE
Don't fallback to series in charm url anymore in bundle deploy

### PR DESCRIPTION
**Pull request by jack-w-shaw**
This is a compatibility layer for a feature we have not supported for a long time

We should not be using the charm url series to determine what OS a charm should be deployed to. This is what the 'base' and 'default-base' options are for

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [ ] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [x] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

### Run integration tests
```
./main.sh -v -c aws -p ec2 deploy test_deploy_bundles
```

### Deploy some common bundles
```
$ juju deploy charmed-kubernetes
```
```
$ juju deploy cos-lite --trust
```
& check applications become active & ready

### Check exported bundles can be deployed correctly
```
$ juju add-model m
$ juju deploy ubuntu ubu1 --base ubuntu@22.04
$ juju deploy ubuntu ubu2 --base ubuntu@20.04
$ juju export-bundle > bundle.bundle
$ cat bundle.bundle
default-base: ubuntu@22.04/stable
applications:
  ubu1:
    charm: ubuntu
    channel: stable
    revision: 24
    num_units: 1
    to:
    - "0"
    constraints: arch=amd64
    storage:
      block: loop,100M
      files: rootfs,100M
  ubu2:
    charm: ubuntu
    channel: stable
    revision: 24
    base: ubuntu@20.04/stable
    num_units: 1
    to:
    - "1"
    constraints: arch=amd64
    storage:
      block: loop,100M
      files: rootfs,100M
machines:
  "0":
    constraints: arch=amd64
  "1":
    constraints: arch=amd64
    base: ubuntu@20.04/stable

$ juju add-model m
$ juju deploy ./bundle.bundle
(wait)
$ juju status
Model  Controller  Cloud/Region         Version    SLA          Timestamp
m2     lxd         localhost/localhost  3.4-beta1  unsupported  13:15:32Z

App   Version  Status   Scale  Charm   Channel  Rev  Exposed  Message
ubu1           waiting      1  ubuntu  stable    24  no       agent initialising
ubu2           waiting      1  ubuntu  stable    24  no       agent initialising

Unit     Workload  Agent      Machine  Public address  Ports  Message
ubu1/0*  waiting   executing  0        10.219.211.10          agent initialising
ubu2/0*  waiting   executing  1        10.219.211.98          agent initialising

Machine  State    Address        Inst id        Base          AZ  Message
0        started  10.219.211.10  juju-812d1f-0  ubuntu@22.04      Running
1        started  10.219.211.98  juju-812d1f-1  ubuntu@20.04      Running
```

### Deploy a bundle without any base/series data in applications
(make sure to include a local charm)

```
$ cat bundle.bundle
applications:
  ubu1:
    charm: ubuntu
    channel: stable
    num_units: 1
    to:
    - "0"
  ubu2:
    charm: /home/jack/charms/ubuntu
    channel: stable
    num_units: 1
    to:
    - "1"
machines:
  "0": {}
  "1":
    base: ubuntu@20.04

$ juju deploy ./bundle.bundle

```

### CODE REVIEW

**Review by wallyworld — APPROVED** [review]
There are no test changes?
A couple of things to look at, but change seems ok to me.
Can we also QA deploying a bundle without a default base in it.

File: core/bundle/changes/handlers.go [hunk 1]
```diff
@@ -1272,44 +1260,30 @@ func applicationKey(charm, arch string, base corebase.Base, channel string, revi
...
 		}
-		return charmURL.Series, nil
+		return base, nil
 	}
-	return "", nil
+
+	return defaultBase, nil
 }
 
 func computeBase(base string, series string, defaultBase corebase.Base) (corebase.Base, error) {
```
[the remark this answers is not in the document]
Comment (jack-w-shaw) [reply core/bundle/changes/handlers.go:1289]: Yes. For machines, or when there isn't a specific application available

File: core/bundle/changes/handlers.go [hunk 2]
```diff
@@ -1272,44 +1260,30 @@ func applicationKey(charm, arch string, base corebase.Base, channel string, revi
...
-// charm path or URL if provided.
-//
-// DEPRECATED: This should be all about bases.
-func getSeries(application *charm.ApplicationSpec, defaultBase corebase.Base) (string, error) {
-	if application.Series != "" {
-		return application.Series, nil
+func computeApplicationBase(application *charm.ApplicationSpec, defaultBase corebase.Base) (corebase.Base, error) {
+	if application.Base != "" {
+		return corebase.ParseBaseFromString(application.Base)
+	} else if application.Series != "" {
```
[the remark this answers is not in the document]
Comment (jack-w-shaw) [reply core/bundle/changes/handlers.go:1266]: I'm not sure if we can remove this by 4.0 as we don't version bundles. But that is defo the long term plan

**Comment by jack-w-shaw** [comment]
No unit test changes, as this should lead to no change in behaviour

I will add QA for a bundle without base or default base or anything now

**Pull request merged**